### PR TITLE
HDDS-7230. Implement GetKeyInfo API

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

Server-side implementation of GetKeyInfo API, which implies 2 primary purposes:
1. It leverages the container location cache to avoid calling SCM to get block location information.
2. It allows S3G to call without a volume name, which the API can assume. S3G can call GetKeyInfo with assumeS3Context and OM will resolve the s3 volume context (volume name and s3 auth) and include it in the API response. This saves S3G another call to get S3VolumeContext before reading a key.

https://issues.apache.org/jira/browse/HDDS-7230

This PR only includes server-side implementation. Full client-side integration and tests will be done by HDDS-7231.
## How was this patch tested?
Unit tests and standard CI: https://github.com/duongkame/ozone/actions/runs/3131964309

### CODE REVIEW

**Comment by duongkame** [comment]
cc. @kerneltime

**Comment by duongkame** [comment]
@adoroszlai It would be great if you can have a look.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java [hunk 1]
```diff
@@ -799,6 +801,34 @@ public OmKeyInfo lookupKey(OmKeyArgs args) throws IOException {
     return OmKeyInfo.getFromProtobuf(resp.getKeyInfo());
   }
 
+  @Override
+  public KeyInfoWithVolumeContext getKeyInfo(OmKeyArgs args,
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java:805]: Would `lookupKeyWithVolumeContext` be more consistent with existing `lookupKey`?  (I assume `lookupKey` has to be retained for compatibility.)

**Review by adoroszlai — COMMENTED** [review]
Thanks @duongkame for the patch, looks mostly good to me.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java [hunk 2]
```diff
@@ -273,6 +273,7 @@ public static boolean isReadOnly(
     case TenantListUser:
     case EchoRPC:
     case RangerBGSync:
+    case GetKeyInfo:
       // RangerBGSync is a read operation in the sense that it doesn't directly
       // write to OM DB. And therefore it doesn't need a OMClientRequest.
       // Although indirectly the Ranger sync service task could invoke write
```
[the remark this answers is not in the document]
Comment (duongkame) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java]: Agree.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyArgs.java [hunk 3]
```diff
@@ -68,12 +70,12 @@ private OmKeyArgs(String volumeName, String bucketName, String keyName,
     this.multipartUploadID = uploadID;
     this.multipartUploadPartNumber = partNumber;
     this.metadata = metadataMap;
-    this.refreshPipeline = refreshPipeline;
```
[the remark this answers is not in the document]
Comment (duongkame) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyArgs.java:71]: Yeah, it's an unused field. I've removed it completely.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/CheckedRunnable.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package org.apache.hadoop.ozone.om.helpers;
+
+import java.io.IOException;
+
+/**
+ * This presents a block of code with a possibility of throwing an IOException.
+ */
+@FunctionalInterface
+public interface CheckedRunnable<E extends IOException> {
```
[the remark this answers is not in the document]
Comment (duongkame) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/CheckedRunnable.java]: Sure thing.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 5]
```diff
@@ -363,10 +367,36 @@ public EncryptedKeyVersion run() throws IOException {
...
+
+      // add block token for read.
+      captureLatencyNs(metrics.getLookupGenerateBlockTokenLatencyNs(),
+          () -> addBlockToken4Read(value));
+
+      // Refresh container pipeline info from SCM
+      // based on OmKeyArgs.refreshPipeline flag
+      // value won't be null as the check is done inside try/catch block.
+      captureLatencyNs(metrics.getLookupRefreshLocationLatencyNs(),
+          () -> refresh(value));
```
[the remark this answers is not in the document]
Comment (duongkame) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:386]: Thanks. Extract a common method to remove the duplication.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java [hunk 6]
```diff
@@ -799,6 +801,34 @@ public OmKeyInfo lookupKey(OmKeyArgs args) throws IOException {
...
+      throws IOException {
+    GetKeyInfoRequest.Builder req = GetKeyInfoRequest.newBuilder();
+    KeyArgs keyArgs = KeyArgs.newBuilder()
+        .setVolumeName(args.getVolumeName())
+        .setBucketName(args.getBucketName())
+        .setKeyName(args.getKeyName())
+        .setDataSize(args.getDataSize())
+        .setSortDatanodes(args.getSortDatanodes())
+        .setLatestVersionLocation(args.getLatestVersionLocation())
+        .setHeadOp(args.isHeadOp())
```
[the remark this answers is not in the document]
Comment (duongkame) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java]: Thanks. I've extracted a common method to remove the duplication.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (duongkame) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:386]: Leveraging the cache also comes with a responsibility to maintain the cached data consistency. In this case, a container location cached in OM may be stale when a relevant datanode is down or the container is moved to another set of datanodes as a result of rebalancing. We've decided to put the responsibility of triggering cache invalidation to the clients when they detect errors reading data from datanodes. This is done by client retrying the new API `getKeyInfo` call with `forceUpdateContainerCacheFromSCM=true`. These interactions are detailed [in this chapter](https://docs.google.com/document/d/1-hVdA7mRBFiFOqr0J0Pg6-Ie5SOZaT7Z/edit#heading=h.4d34og8).

For backward compatibility with previous client versions, we need to keep `lookupKey` as it is.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java (hunk 1 above)
Comment (duongkame) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java:796]: It's fair. Yet I think we're trying to deprecate the `lookup` APi and replace it with an optimized API. The name `get` sounds more appropriate as this API takes a key identifier (vol/bucket/keyname) and returns KeyInfo and it does zero lookup/search. 
How do you think @kerneltime?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:386]: Thanks for the explanation.

**Review by adoroszlai — APPROVED** [review]

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java (hunk 1 above)
Comment (kerneltime) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OzoneManagerProtocolClientSideTranslatorPB.java:796]: Lookup is a very specific API in the context of filesystems, and this API is nothing like it. It is more akin to get attributes of a file.

**Review by kerneltime — APPROVED** [review]

**Comment by kerneltime** [comment]
I can merge this in once conflicts are resolved.

**Comment by duongkame** [comment]
> I can merge this in once conflicts are resolved.

Thanks for the review @kerneltime, @adoroszlai . 

I've just resolved the conflicts. Let's see if there's any problem with CI workflow.

**Comment by adoroszlai** [comment]
Thanks @duongkame for the patch, @kerneltime for the review.

**Pull request merged**